alerta/auth/basic_ldap.py

```diff
@@ -0,0 +1,72 @@
+
+import logging
+from uuid import uuid4
+
+from flask import current_app, request, jsonify, render_template
+from flask_cors import cross_origin
+
+from alerta.auth.utils import is_authorized, create_token, get_customers
+from alerta.exceptions import ApiError
+from alerta.models.user import User
+from alerta.utils.api import absolute_url
+from . import auth
+
+import ldap
+
+@auth.route('/auth/signup', methods=['OPTIONS', 'POST'])
+@cross_origin(supports_credentials=True)
+def signup():
+    raise NotImplementedError
+
+@auth.route('/auth/login', methods=['OPTIONS', 'POST'])
+@cross_origin(supports_credentials=True)
+def login():
+    # Retrieve required fields from client request
+    try:
+        email = request.json.get('username', None) or request.json['email']
+        password = request.json['password']
+    except KeyError:
+        raise ApiError("must supply 'username' and 'password'", 401)
+
+    username = email.split("@")[0]
+    domain = email.split("@")[1]
+
+    # Validate LDAP domain
+    if not 'LDAP_DOMAINS' in current_app.config:
+        raise ApiError("LDAP_DOMAINS not configured", 500)
+
+    if domain not in current_app.config["LDAP_DOMAINS"]:
+        raise ApiError("unauthorized domain", 403)
+
+    userdn = current_app.config["LDAP_DOMAINS"][domain] % username
+
+    # Attempt LDAP AUTH
+    try:
+        ldap_connection = ldap.initialize(current_app.config['LDAP_SERVER'])
+        ldap_connection.simple_bind_s(userdn, password)
+
+    except ldap.INVALID_CREDENTIALS:
+        raise ApiError("invalid username or password", 401)
+
+    except:
+        raise ApiError("logon failed", 500)
+
+    # Create user if not yet there
+    user = User.find_by_email(email=email)
+    if not user:
+        user = User(username, email, "", ["user"], "LDAP user")
+        user.create()
```

Comment:
Why are you creating the user in the Alerta database when it already exists in the LDAP database? What purpose does that serve?

Reply:
Hi!

This is to keep using Alerta's framework for authorization, only the authentication is offloaded to LDAP. 

Users, permissions and roles are still controlled by Alerta here. 

By the way, a colleague of mine will get back to you regarding the question regarding dev LDAP server.

Reply:
I would have thought it would make sense to make full use of the LDAP database and not just use it for authentication. The LDAP server could also be used to set user roles by mapping LDAP groups to Alerta roles. Similar to how it done for the Keycloak and SAML integrations.

See these examples of how (unfortunately complicated) it can be done... http://docs.grafana.org/installation/ldap/ and https://access.redhat.com/documentation/en-us/red_hat_jboss_operations_network/3.1/html/admin_initial_setup_inventory_groups_and_users/about-ldap-groups

As a first iteration, it might be good enough to just use it for authentication and add further integration at a later date.

Reply:
Hmm, I agree. 
For our usecase it might be that we require LDAP groups as well, I need to check up on that. It's very possible that you'll get another pull request for implementing LDAP groups too in the very near future